### PR TITLE
JUnit dependencies in MockJoiner and TestNodeRegistry

### PR DESCRIPTION
**Pull request by rvanheest**
This PR is based on issue #8868. It removes the dependency on JUnit for the `MockJoiner` and `TestNodeRegistry` classes.

### CODE REVIEW

**Comment by Donnerbart** [comment]
run-lab-run

**Comment by jerrinot** [comment]
Hi @rvanheest,

thanks for this contribution. Relying on JVM-level assertions being enabled is a tricky business. I would rather go with an explicit check and throwing `AssertionError`. WDYT?

**Comment by rvanheest** [comment]
@jerrinot indeed this is a little tricky. However, if you want to through these `AssertionError`s instead, you need to do the if-check there as well, which is already done for you in the `assert`. Being more of a Scala guy myself, I know there are special functions such as [`assert()`](http://www.scala-lang.org/api/current/index.html#scala.Predef$@assert(assertion:Boolean,message:=>Any):Unit) in that language that would do the trick. I could create something similar for these changes instead if you want... but if you think an if-check by hand would suffice, I'd be more than happy to change it in that way!

Let me know!

**Comment by jerrinot** [comment]
@rvanheest: I understand. The problem is the `assert`  keyword is checked only when a JVM is started with the `-ea` flag. I don't think we should rely on it. 

We have had a quite few issues with the `MockJoiner` in the past - they can be pretty nasty and hard-to-find -> We should make sure invariants are always checked, no matter what flags were used to start the JVM.

**Comment by rvanheest** [comment]
absolutely agree on that! Those invariants should be checked at all time, no matter the flags. I'll change the `assert` keywords to 'normal' checks and `AssertionError`s.

**Comment by rvanheest** [comment]
@jerrinot See this new commit (ffaf1c6). I created a static method called `verifyInvariant(boolean check, String msg)` that throws an `AssertionError` with the `msg` if `check` is false. For now I included this as a private function to both files, as you probably don't want to expose this function in the API. However, I realize that this is code duplication, which is probably not desirable. Any suggestions as to where I could best put this function?

**Comment by rvanheest** [comment]
(your response in [the issue](https://github.com/hazelcast/hazelcast/issues/8868#issuecomment-246659026) 😉 )

Missed that `HazelcastTestSupport` class. Nice trick to have a abstract class for this! I added `verifyInvariant` there and did an `extend` in `TestNodeRegistry`. However, as `MockJoiner` already extends another class I wasn't able to do so there. Instead I used static importing there.

File: hazelcast/src/test/java/com/hazelcast/test/mocknetwork/TestNodeRegistry.java [hunk 1]
```diff
@@ -32,9 +33,7 @@
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
 
-import static org.junit.Assert.assertEquals;
-
-public final class TestNodeRegistry {
+public final class TestNodeRegistry extends HazelcastTestSupport {
```
Comment (jerrinot) [inline hazelcast/src/test/java/com/hazelcast/test/mocknetwork/TestNodeRegistry.java]: it's better to use static imports here as well.

**Comment by jerrinot** [comment]
:+1:

File: hazelcast/src/test/java/com/hazelcast/test/mocknetwork/TestNodeRegistry.java (hunk 1 above)
Comment (rvanheest) [inline hazelcast/src/test/java/com/hazelcast/test/mocknetwork/TestNodeRegistry.java]: Ah, I see. I saw this construct in some other places, so I thought I might copy it. But this is better indeed.

**Comment by jerrinot** [comment]
@rvanheest: looks good to me. Have you signed Hazelcast Contributor Agreement? Your name rings same bell, but I don't see you at https://hazelcast.atlassian.net/wiki/display/COM/Hazelcast+Contributor+Agreement

**Comment by rvanheest** [comment]
Just sent it.

**Comment by Donnerbart** [comment]
run-lab-run

**Comment by Donnerbart** [comment]
Hmm, `HazelcastTestSupport` has dependencies on `org.junit.Assert` again. Don't we introduce the dependency the changed classes again?

**Comment by jerrinot** [comment]
we don't. it's fine as long as you don't call a method depending on it. (or do stuff like proxying, etc..)

**Comment by rvanheest** [comment]
Nope, it's not fine! I'm looking into it, but I overlooked the whole purpose of this PR for a second. Let me look into it a bit.

**Comment by Donnerbart** [comment]
:+1: (once green)

**Comment by rvanheest** [comment]
OK, figured it out: since we do a static import, `XXX()` is on compile time rewritten to `HazelcastTestSupport.XXX()`. There it still requires all imports of the `HazelcastTestSupport` class, which indeed contain JUnit imports...

I now rewrote it to a private static method in both classes again. I think that is the safest way for now. I considered creating a new class for this, but it didn't feel appropriate to have a class with only one static method in it next to the `HazelcastSupport` class. Would be too risky for future development to put this function back in the other class...

**Comment by Donnerbart** [comment]
run-lab-run

**Comment by rvanheest** [comment]
Hi, I was wondering what's up with this PR. Do I still need to do something before it can be merged?

**Comment by jerrinot** [comment]
@rvanheest: thanks for the kick!

**Pull request merged**